src/kernel/syscalls.zig

```diff
@@ -97,6 +181,122 @@ pub fn handle(syscall: Syscall, arg1: usize, arg2: usize, arg3: usize, arg4: usi
     return try syscall.getHandler()(arg1, arg2, arg3, arg4, arg5);
 }
 
+///
+/// Get a slice containing the data at an address and length. If the current task is a kernel task then a simple pointer to slice conversion is performed,
+/// otherwise the slice is allocated on the heap and the data is copied in from user space.
+///
+/// Arguments:
+///     IN ptr: usize - The slice's address
+///     IN len: usize - The number of bytes
+///
+/// Error: Error
+///     Error.OutOfMemory - There wasn't enough kernel (heap or VMM) memory left to fulfill the request.
+///     Error.TooBig - The user task requested to have too much data copied
+///     Error.InvalidAddress - The address that the user task passed is invalid (not mapped, out of bounds etc.)
+///
+/// Return: []u8
+///     The slice of data. Will be stack-allocated if the current task is kernel-level, otherwise will be heap-allocated
+///
+fn getData(ptr: usize, len: usize) Error![]u8 {
+    if (scheduler.current_task.kernel) {
+        return @intToPtr([*]u8, ptr)[0..len];
+    } else {
+        if (len > USER_MAX_DATA_LEN) {
+            return Error.TooBig;
+        }
+        var buff = try allocator.alloc(u8, len);
+        errdefer allocator.free(buff);
+        try vmm.kernel_vmm.copyData(scheduler.current_task.vmm, buff, ptr, false);
+        return buff;
+    }
+}
+
+fn handleOpen(path_ptr: usize, path_len: usize, flags: usize, args: usize, ignored: usize) Error!usize {
+    const current_task = scheduler.current_task;
+    if (!current_task.hasFreeVFSHandle()) {
+        return Error.NoMoreFSHandles;
+    }
+
+    // Fetch the open arguments from user/kernel memory
+    var open_args: vfs.OpenArgs = if (args == 0) .{} else blk: {
```

Comment:
const

Reply:
This can't be const unfortunately since the `symlink_target` fields needs to be modified after setting this.